mypy/traverser.py

```diff
@@ -352,6 +355,476 @@ def visit_exec_stmt(self, o: ExecStmt) -> None:
+
+    def visit_enum_call_expr(self, o: EnumCallExpr) -> None:
+        if not self.visit(o):
+            return
+        super().visit_enum_call_expr(o)
+
+    def visit_typeddict_expr(self, o: TypedDictExpr) -> None:
+        if not self.visit(o):
+            return
+        super().visit_typeddict_expr(o)
+
+    def visit_newtype_expr(self, o: NewTypeExpr) -> None:
+        if not self.visit(o):
+            return
+        super().visit_newtype_expr(o)
+
+    def visit_await_expr(self, o: AwaitExpr) -> None:
+        if not self.visit(o):
+            return
+        super().visit_await_expr(o)
+
+    # Patterns
+
+    def visit_as_pattern(self, o: AsPattern) -> None:
+        if not self.visit(o):
+            return
+        super().visit_as_pattern(o)
+
+    def visit_or_pattern(self, o: OrPattern) -> None:
+        if not self.visit(o):
+            return
+        super().visit_or_pattern(o)
+
+    def visit_value_pattern(self, o: ValuePattern) -> None:
+        if not self.visit(o):
+            return
+        super().visit_value_pattern(o)
+
+    def visit_singleton_pattern(self, o: SingletonPattern) -> None:
+        if not self.visit(o):
+            return
+        super().visit_singleton_pattern(o)
+
+    def visit_sequence_pattern(self, o: SequencePattern) -> None:
+        if not self.visit(o):
+            return
+        super().visit_sequence_pattern(o)
+
+    def visit_starred_pattern(self, o: StarredPattern) -> None:
+        if not self.visit(o):
+            return
+        super().visit_starred_pattern(o)
+
+    def visit_mapping_pattern(self, o: MappingPattern) -> None:
+        if not self.visit(o):
+            return
+        super().visit_mapping_pattern(o)
+
+    def visit_class_pattern(self, o: ClassPattern) -> None:
+        if not self.visit(o):
+            return
+        super().visit_class_pattern(o)
+
+
+class SearchVisitor(ExtendedTraverserVisitor):
+    def __init__(
+        self,
+        line: int,
+        column: int,
+        end_line: int,
+        end_column: int
+    ) -> None:
+        self.line = line
+        self.column = column
+        self.end_line = end_line
+        self.end_column = end_column
+        self.result: Optional[Expression] = None
+
+    def visit(self, o: Node) -> bool:
+        if o.line > self.line or o.line == self.line and o.column > self.column:
+            return False
+        # Unfortunately, end positions for some statements are a mess,
+        # e.g. overloaded functions, so we can't optimise by skipping them.
+        if o.end_line is not None and o.end_column is not None:
+            if (o.end_line < self.end_line
+                    or o.end_line == self.end_line and o.end_column < self.end_column):
+                return False
+        if (
+            o.line == self.line
+            and o.end_line == self.end_line
+            and o.column == self.column
+            and o.end_column == self.end_column
+        ):
```

Comment:
Requiring an exact match of the span here can make it hard to implement integrations. E.g., consider an editor extension that shows the type of an expression via a command in the right click menu. Now to get the correct span, the editor should probably use the Python parser, and maybe even the same version of the parser that is used by mypy, since I assume the spans could vary slightly between Python versions. So basically the editor would need to implement a similar visitor to find the span, which seems too complicated, especially if the editor is written in, say, Java or TypeScript.

Some ideas about how to deal with this below.

First, we could only take (line, column) as input, and return types for all the expressions that span this location. So if we have `x = foo + bar` and the user requests the type of the location in the middle of `foo`, we might return the type of `foo` and the type of `foo + bar` (with the corresponding spans). The caller can then decide what to do with the results: for example, they could show the innermost type only, or they could show all of the types (or the N innermost types), or they could ask the user to pick with expression they are interested in as a second step.

Second, we could always return the type for the innermost expression that includes the location (or intersects the span, if we take end line/column as well). This would means that to get the type of a binary operation, the location would have to be the between the operands (e.g. the operator). Here is an example (^ indicates the location requested by the caller):
```
foo + bar
 ^                return the type of "foo"
foo + bar
      ^           return the type of "bar"
foo + bar
    ^             return the type of "foo + bar"
```

Reply:
Hm, I think we can accept both forms, so that if user gives us full span (4 numbers), we return just the exact match (e.g. if user uses mouse to select a span). But if the user gives only 2 (i.e. line/column) we return the full stack starting from innermost (e.g. if user uses a right click).

I think we can also have flags to:
* Include span in response
* Include expression kind in response (maybe people want specifically `CallExpr`, not the argument etc)
* Always include only innermost expression (for 2 number requests)